### ISSUE TITLE
cannot construct populate class with Ref() 

### ISSUE THREAD

**Issue by nhhockeyplayer** [typegoose/typegoose#967]
Unsure of this is bug or design limitation but Im kinda stuck and hoping for some clarification, workaround... help
thanks in advance

Im building against a privately packed/installed typegoose feature/esm library

cloned typegoose feature/esm branch
added script to package.json
```json
  "scripts": {
    "pack-esm": "cd ./esm && npm pack @typegoose/typegoose@12.9.1",
```

npm pack produced for my package.json
`        "@typegoose/typegoose": "file:typegoose-typegoose-12.9.1.tgz",`

and I am changing all my models back to typegoose compliant

I am stuck on a life-cycle issue class instance creation/population

## What is the Problem?
I cannot create the code to instantiate my class instance

Here is my typegoose model

```ts
export class UserPojo extends AbstractPojo<UserPojo> implements IUserPojo, OmitFunctions<UserPojo> {
    @prop()
    emailAddress!: string
    @prop()
    password?: string

    @prop({ ref: () => RolePojo })
    role?: Ref<RolePojo>

    constructor(anatomy: Partial<UserPojo> = {role: null}) {
        super(anatomy)
    }

    override hash?(): Types.ObjectId | string {
        return this._id === undefined ? '' : this._id
    }
}
```

now normally before I decided to re-convert back to typegoose models I could do the following.

    /* create new userpojo with a new ref to rolepojo */
    user: UserPojo = new UserPojo({
        role: new RolePojo({
            name: ERole.ADMINISTRATOR,
            description: ERole.ADMINISTRATOR,
            isSuper: true
        }),
        password: 'jdoe',
        emailAddress: 'jdoe@gmail.com'
    })

But since now Im using typegoose we have Ref() property/handle to a class instance now

```ts
    @prop({ ref: () => RolePojo })
    role?: Ref<RolePojo>
```

these lines are at the bottom of the class but they seem kinda useless to me I dont know how they would fit in
```ts
export const UserPojoModel = getModelForClass(UserPojo);
export const docUserPojo = new UserPojoModel();
```

and I do not know how to get past this compiler issue to create a class model instance on the fly as shown prior with straight typescript with no typegoose

actually I don't know how to do it due to compiler errors

can someone clarify and write the code to create an instance for this?

should I just create one like this?
```ts
    userPojo = new UserPojo({
        _id: null
    })
```
or like this?

`    documentPojo = new UserPojoModel(this.userPojo)`

it seems Im unable to even just spread properties and objects because of the Ref() type

Is it possible to even do a NEW on the Ref() class type?

In short I have properties of my model UserPojo that are references to other models RolePojo
and I want to know how to create both models on the fly by instantiating the containing model

sorry if this sounds confusing
but Im not seeing flexibility here like I do with plain interfaces and spread operator to populate the properties of a property that points to another model

here is the actual compiler error Im receiving
```
 NX   libs/shared/root/security/auth/src/lib/guard/authenticated-guard.service.ts:20:7 - error TS2322: Type 'RolePojo' is not assignable to type 'Ref<RolePojo>'.
20       role: new RolePojo({
```

liine 20 looks like this

```ts
    user: UserPojo = new UserPojo({
      role: new RolePojo({
        name: ERole.ADMINISTRATOR,
        description: ERole.ADMINISTRATOR,
        isSuper: true
      }),
      password: 'jdoe',
      emailAddress: 'jdoe@gmail.com'
    })
```

so this is the typescript inspired way to whip off a model on the fly

but Im lost with regard to how to handle Ref()

thanks

---
NX REPORT

```
% nx report

 NX   Report complete - copy this into the issue template

Node           : 22.9.0
OS             : darwin-x64
Native Target  : x86_64-macos
npm            : 10.9.0

nx (global)        : 20.0.0
nx                 : 20.2.1
@nx/js             : 20.2.1
@nx/jest           : 20.2.1
@nx/eslint         : 20.2.1
@nx/workspace      : 20.2.1
@nx/angular        : 20.2.1
@nx/cypress        : 20.2.1
@nx/devkit         : 20.2.1
@nx/esbuild        : 20.2.1
@nx/eslint-plugin  : 20.2.1
@nx/express        : 20.2.1
@nx/nest           : 20.2.1
@nx/node           : 20.2.1
@nx/playwright     : 20.2.1
@nx/vite           : 20.2.1
@nx/web            : 20.2.1
@nx/webpack        : 20.2.1
nx-cloud           : 19.1.0
typescript         : 5.6.3
---------------------------------------
Registered Plugins:
@nx/cypress/plugin
@nx/eslint/plugin
@nx/playwright/plugin
---------------------------------------
Community plugins:
@compodoc/compodoc               : 1.1.26
@fortawesome/angular-fontawesome : 1.0.0
@ionic/angular                   : 6.7.5
@jsverse/transloco               : 7.5.0
@ngrx/component                  : 18.0.2
@ngrx/component-store            : 18.0.2
@ngrx/effects                    : 18.0.2
@ngrx/entity                     : 18.0.2
@ngrx/operators                  : 18.0.2
@ngrx/router-store               : 18.0.2
@ngrx/schematics                 : 18.0.2
@ngrx/store                      : 18.0.2
@ngrx/store-devtools             : 18.0.2
@nxext/capacitor                 : 20.1.0
@nxext/ionic-angular             : 20.0.5
apollo-angular                   : 8.0.0
```

**Comment by nhhockeyplayer** [comment]
```
import { AbstractPojo } from './abstract/abstract.pojo'
import { IUserPojo } from './interfaces/i-user.pojo'
import { Ref } from 'react'
import { RolePojo } from './role.pojo'
import { OmitFunctions } from './interfaces/i-business.pojo'
import { EncipheredTokensPojo } from './enciphered-tokens.pojo'
import { getModelForClass, prop } from '@typegoose/typegoose'
import mongoose, { Types } from 'mongoose'

export class UserPojo extends AbstractPojo<UserPojo> implements IUserPojo, OmitFunctions<UserPojo> {
    @prop()
    emailAddress!: string
    @prop()
    password?: string

    @prop({ ref: () => EncipheredTokensPojo })
    tokens?: Ref<EncipheredTokensPojo>

    @prop({ ref: () => RolePojo })
    role?: Ref<RolePojo>

    @prop()
    content?: string[] | Buffer | null | undefined
    @prop()
    contentType?: string

    constructor(anatomy: Partial<UserPojo> = {role: null}) {
        super(anatomy)
    }

    override hash?(): Types.ObjectId | string {
        return this._id === undefined ? '' : this._id
    }
}

export const UserPojoModel = getModelForClass(UserPojo);

export const docUserPojo = new UserPojoModel();
```

**Comment by nhhockeyplayer** [comment]
The enciphered tokens ref is even more complicated and deeper nested

```
export class EncipheredTokensPojo extends AbstractPojo<EncipheredTokensPojo> implements IEncipheredTokens, OmitFunctions<EncipheredTokensPojo> {
    @prop({ ref: () => TokenPojo })
    accessToken: Ref<TokenPojo>
    @prop({ ref: () => TokenPojo })
    refreshToken: Ref<TokenPojo>

    constructor(anatomy: Partial<EncipheredTokensPojo> = {}) {
        super(anatomy)
    }

    hash?() {
        return this._id === undefined ? '' : this._id
    }
}

export const EncipheredTokensPojoModel = getModelForClass(EncipheredTokensPojo)
```

and the token

```
export class TokenPojo extends AbstractPojo<TokenPojo> implements IToken, OmitFunctions<TokenPojo> {
    @prop()
    token: string
    @prop()
    expiresAt: string

    constructor(anatomy: Partial<TokenPojo> = {}) {
        super(anatomy)
    }

    override hash?() {
        return this._id === undefined ? '' : this._id
    }
}

export const TokenPojoModel = getModelForClass(TokenPojo)

```

**Comment by nhhockeyplayer** [comment]
heres a good example of my issue

```
    public login(
        emailAddress: string,
        password: string
    ): Observable<UserPojo> {
        if (emailAddress === 'wu' && password === 'wp') {
            const expiresAtInt = Date.now() + 300000 // 5 minutes later
            const expiresAt = new Date(expiresAtInt).toISOString()
            const emailAddress = 'meanstack@hotmail.com'

            const accessToken = new TokenPojo({
                token: 'dummy token',
                expiresAt
            })

            const refreshToken = new TokenPojo({
                token: 'dummy token',
                expiresAt
            })

            const user: UserPojo = new UserPojo({
                tokens: new EncipheredTokensPojo({
                    accessToken: accessToken,
                    refreshToken: refreshToken,
                }),
                emailAddress
            })
            return of(user)
        } else {
            return of(null)
    }

```

and the error on properties tokens, accessToken, refreshToken key/properties

thanks for any assist
```
TS2322: Type EncipheredTokensPojo is not assignable to type Ref<EncipheredTokensPojo>
user.pojo.ts(17, 5): The expected type comes from property tokens which is declared here on type Partial<UserPojo>
```

**Comment by nhhockeyplayer** [comment]
let me know if you need me to post any more code

thanks for any feedback

- cheers

**Comment by hasezoey** [comment]
For easier reading of provided code, please add language specifiers to [enable syntax highlighting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting).

---

Type `Ref` is a helper type that converts a given [`Class` to a Mongoose Document](https://github.com/typegoose/typegoose/blob/d775d94f48adaa058f3428b59738610fee158e07/src/types.ts#L417-L422) and also ORs it with the reference type (like objectid), so you can only assign mongoose documents to that property.
In short you would need to replace your plain class creations (like `new TokenPojo`) with the document creations (like `new TokenPojoModel`).
Alternatively, you could also just add a OR to the class itself, if that is how you want to keep it. Like `Ref<Class> | Class`. Though note that you might be forced to check for types when trying to access the type.

**Comment by nhhockeyplayer** [comment]
thanks friend
can you post a runnable snippet?
close to my last post
to convert data into an object like I tried

**Comment by hasezoey** [comment]
>  can you post a runnable snippet?

Here is a example of turning all your calls to documents:

```ts
/* eslint-disable @typescript-eslint/no-unused-vars */
// NodeJS: 22.8.0
// MongoDB: 7.0 (Docker)
// Typescript 5.3.3
import { DocumentType } from '@typegoose/typegoose';
import { getModelForClass, prop, Ref } from '@typegoose/typegoose'; // @typegoose/typegoose@12.9.0
import * as mongoose from 'mongoose'; // mongoose@8.8.0

class TokenPojo {
  @prop()
  public token?: string;

  @prop()
  public expiresAt?: string;

  constructor(anatomy: Partial<TokenPojo> = {}) {}

  hash?(this: DocumentType<TokenPojo>) {
    return this._id === undefined ? '' : this._id;
  }
}

const TokenPojoModel = getModelForClass(TokenPojo);

class EncipheredTokensPojo {
  @prop({ ref: () => TokenPojo })
  public accessToken?: Ref<TokenPojo>;

  @prop({ ref: () => TokenPojo })
  public refreshToken?: Ref<TokenPojo>;

  constructor(anatomy: Partial<EncipheredTokensPojo> = {}) {}

  hash?(this: DocumentType<TokenPojo>) {
    return this._id === undefined ? '' : this._id;
  }
}

export const EncipheredTokensPojoModel = getModelForClass(EncipheredTokensPojo);

class UserPojo {
  @prop({ ref: () => EncipheredTokensPojo })
  public tokens?: Ref<EncipheredTokensPojo>;

  constructor(anatomy: Partial<UserPojo> = {}) {}

  hash?(this: DocumentType<TokenPojo>): mongoose.Types.ObjectId | string {
    return this._id === undefined ? '' : this._id;
  }
}

const UserPojoModel = getModelForClass(UserPojo);

async function main() {
  await mongoose.connect(`mongodb://localhost:27017/`, {
    dbName: 'verifyMASTER',
  });

  const expiresAtInt = Date.now() + 300000; // 5 minutes later
  const expiresAt = new Date(expiresAtInt).toISOString();

  const accessToken = new TokenPojoModel({
    token: 'dummy token',
    expiresAt,
  });

  const refreshToken = new TokenPojoModel({
    token: 'dummy token',
    expiresAt,
  });

  await Promise.all([accessToken.save(), refreshToken.save()]);

  const user = new UserPojoModel({
    tokens: new EncipheredTokensPojoModel({
      accessToken: accessToken,
      refreshToken: refreshToken,
    }),
  });

  await (user.tokens as DocumentType<EncipheredTokensPojo>).save();
  await user.save();

  await mongoose.disconnect();
}

main();
```
(commit name `showcase 1`)

and a example for adding `| class`:

```ts
/* eslint-disable @typescript-eslint/no-unused-vars */
// NodeJS: 22.8.0
// MongoDB: 7.0 (Docker)
// Typescript 5.3.3
import { DocumentType } from '@typegoose/typegoose';
import { getModelForClass, prop, Ref } from '@typegoose/typegoose'; // @typegoose/typegoose@12.9.0
import * as mongoose from 'mongoose'; // mongoose@8.8.0

class TokenPojo {
  @prop()
  public token?: string;

  @prop()
  public expiresAt?: string;

  constructor(anatomy: Partial<TokenPojo> = {}) {
    for (const [key, val] of Object.entries(anatomy)) {
      // @ts-expect-error This is just for quick assignment showcase
      this[key] = val;
    }
  }

  hash?(this: DocumentType<TokenPojo>) {
    return this._id === undefined ? '' : this._id;
  }
}

const TokenPojoModel = getModelForClass(TokenPojo);

class EncipheredTokensPojo {
  @prop({ ref: () => TokenPojo })
  public accessToken?: Ref<TokenPojo> | TokenPojo;

  @prop({ ref: () => TokenPojo })
  public refreshToken?: Ref<TokenPojo> | TokenPojo;

  constructor(anatomy: Partial<EncipheredTokensPojo> = {}) {
    for (const [key, val] of Object.entries(anatomy)) {
      // @ts-expect-error This is just for quick assignment showcase
      this[key] = val;
    }
  }

  hash?(this: DocumentType<TokenPojo>) {
    return this._id === undefined ? '' : this._id;
  }
}

export const EncipheredTokensPojoModel = getModelForClass(EncipheredTokensPojo);

class UserPojo {
  @prop({ ref: () => EncipheredTokensPojo })
  public tokens?: Ref<EncipheredTokensPojo> | EncipheredTokensPojo;

  constructor(anatomy: Partial<UserPojo> = {}) {
    for (const [key, val] of Object.entries(anatomy)) {
      // @ts-expect-error This is just for quick assignment showcase
      this[key] = val;
    }
  }

  hash?(this: DocumentType<TokenPojo>): mongoose.Types.ObjectId | string {
    return this._id === undefined ? '' : this._id;
  }
}

const UserPojoModel = getModelForClass(UserPojo);

async function main() {
  await mongoose.connect(`mongodb://localhost:27017/`, {
    dbName: 'verifyMASTER',
  });

  const expiresAtInt = Date.now() + 300000; // 5 minutes later
  const expiresAt = new Date(expiresAtInt).toISOString();

  const accessToken = new TokenPojo({
    token: 'dummy token',
    expiresAt,
  });

  const refreshToken = new TokenPojo({
    token: 'dummy token',
    expiresAt,
  });

  const user = new UserPojo({
    tokens: new EncipheredTokensPojo({
      accessToken: accessToken,
      refreshToken: refreshToken,
    }),
  });

  // all classes are now actual class instances, not documents
  console.log('user', user);

  await mongoose.disconnect();
}

main();
```
(commit name `showcase 2`)
(Notes: disregard the `hash` function `this` typing, it wasnt part of this question, it would be more correct to also add `| Class` there, but i didnt want to add the unprovided `AbstractPojo`.
Also note that the assignment of values should likely be checked more than this "ts complaining quick and dirty way".)

Both examples can be found (under the specified commit names) in the [`typegoose-testing` repository at branch `verify967`](https://github.com/typegoose/typegoose-testing/tree/verify967).

**Comment by nhhockeyplayer** [comment]
I will try to digest and work with this thank you
but in all fairness
this does deserve an update to the docs in a section on life-cycle (instantiation)
most formidable developers object oriented use polymorphism and I do typescript does everything java does I got 16 yrs hard java r&d
constructing an instance of a taxonomy (class hierarchy) tied together thru Ref() should be clean using a composition of objects and properties mine are processed thru my copy constructors

as depicted here all object sand properties
```
            const accessToken = new TokenPojo({
                token: 'dummy token',
                expiresAt
            })

            const refreshToken = new TokenPojo({
                token: 'dummy token',
                expiresAt
            })

            const user: UserPojo = new UserPojo({
                tokens: new EncipheredTokensPojo({
                    accessToken: accessToken,
                    refreshToken: refreshToken,
                }),
                emailAddress
            })
```

this will tie up loose ends with model integrity and relationships

I will see this thru and close out next week sometime with my results

but if some working examples short concise can be placed in the docs that would be terrific

I feel good about typegoose now and hope to land a back end powered by m8 on a nestJS platform

thanks

**Comment by nhhockeyplayer** [comment]
let me know if I can provide you with my base class taxonomy
its pretty elite

**Comment by nhhockeyplayer** [comment]
well Im seeing flakey corruption results combining m8 and typegoose esm

m8 is anchored to an earlier version of typegoose

so settling on the esm branch is futile for me there are collisions

1. is typegoose for esm currently releasing as main package from github now?
2. can m8 synch up with that?

Thanks for any update or path forward to get this in synch

Appreciate any advice for path forwaRd

**Comment by smolinari** [comment]
@nhhockeyplayer 

When you say m8, what are you referring to? The m8a nestjs-typegoose module? 

If yes, write an issue in my repository. And please be more specific than "flaky corruption results". I can't work with that. And for sure, and maybe this will answer your question, I'm not releasing an ESM version of nestjs-typegoose any time soon. 

Scott

**Comment by hasezoey** [comment]
> well Im seeing flakey corruption results combining m8 and typegoose esm
> m8 is anchored to an earlier version of typegoose

This is unrelated to the current issue, so i am marking it off-topic.

> is typegoose for esm currently releasing as main package from github now?

No, typegoose is currently not publishing any ESModules currently, they are only experimentally available in branch `feature/esm`.
I will also not publish this code until it is at least somewhat tested (even if it requires flags), but even if i do, as long as it requires workarounds (like extra node flags) it will be experimental.

And once again, even if typegoose is ESM, mongoose and mongodb are not currently.

PS: please keep ESM talk in either #214 or a new discussion.

**Comment by hasezoey** [comment]
Considering that there have not been any new questions regarding the original issues question, i will close this.

**Comment by nhhockeyplayer** [comment]
my questions were off topic? trying to see this thru?
someone having a bad day somewhere?

**Comment by hasezoey** [comment]
> my questions were off topic? trying to see this thru?

From what i can tell, this issue (#967), is (or at least was) about "assigning values to Ref\<Class> fields", not about "m8 and/or esmodules".
(yes they were mentioned in the Original post, but not relevant to the question from what i can tell and your new comment not seeming relevant to failing because of the Ref\<Class> changes)

Also, why where are your previous comments? This kinda makes this thread hard to follow now.

**Comment by nhhockeyplayer** [comment]
Im fins closing this.
Understood.
Can you answer one question?
Did you merge the esm Branch with master (should be main but thats okay)?
I am encountering corruption and since Im tied to m8 which is tied to early version of typegoose I can workaround this by synching to latest release if in fact its merged.
Thanks

**Comment by hasezoey** [comment]
> Did you merge the esm Branch with master

No, as stated [in this comment](https://github.com/typegoose/typegoose/issues/967#issuecomment-2558123014) and [this comment](https://github.com/typegoose/typegoose/issues/214#issuecomment-2543173168):
- the `feature/esm` branch had been updated to typegoose 12.10.0
-  some changes from `feature/esm` had been upstreamed
- the `feature/esm` branch is **not** merged into `master`
- it might be possible to merge the 2 branches, but either without publishing the ESM side, or with very clear experimental notes.

(though if i would do merge / publish it, it would likely only be on next major version)

**Comment by nhhockeyplayer** [comment]
thank you
looking forward to -->merge / publish it, it would likely only be on next major version
TypeGoose 13.0.0, should be a breakout I will definitely demonstrate a commercial usage
